### ISSUE TITLE
Use json.RawMessage to represent `JSON` type struct fields instead of []byte

### ISSUE THREAD

**Issue by Goldziher** [sqlc-dev/sqlc#2965]
### What do you want to change?

Hi there,

I'd like to ask for an update to how structs are generated. Instead of using `[]byte` to represent a `JSON` type field, use `json.RawMessage`. This affects serialization and deserialization primarily.  It could of course just be optional. 

### What database engines need to be changed?

PostgreSQL

### What programming language backends need to be changed?

Go

**Comment by andrewmbenton** [comment]
As far as I can tell, sqlc does use `json.RawMessage` for PostgreSQL `JSON` database columns. Here's a playground link to demonstrate: https://play.sqlc.dev/p/e54dc46953ed8e3c1be0875c17b84ddb53f65825afc1f20b014d78309331f5b2

In what context are you seeing `[]byte`?

**Comment by Goldziher** [comment]
> As far as I can tell, sqlc does use `json.RawMessage` for PostgreSQL `JSON` database columns. Here's a playground link to demonstrate: https://play.sqlc.dev/p/e54dc46953ed8e3c1be0875c17b84ddb53f65825afc1f20b014d78309331f5b2
> 
> In what context are you seeing `[]byte`?

it does not use it to represent struct fields or on receivers for me.

**Comment by andrewmbenton** [comment]
> it does not use it to represent struct fields or on receivers for me.

Understood, I just can't reproduce that behavior. Can you provide an example using https://play.sqlc.dev or modify the one I linked previously?

**Comment by Goldziher** [comment]
> > it does not use it to represent struct fields or on receivers for me.
> 
> Understood, I just can't reproduce that behavior. Can you provide an example using https://play.sqlc.dev or modify the one I linked previously?

You are very nice! thanks. 

The reproduction is simple, the issue is when using `pgx`. 

See this: https://play.sqlc.dev/p/02d8afb0d4d5c2f12f31141833558cb6342747a6eb44f6264f2c2462d2f276b1

I updated the config to use pgx, and added a json value to the model. The result is this:

[
<img width="1728" alt="Screenshot 2023-11-18 at 18 08 55" src="https://github.com/sqlc-dev/sqlc/assets/30733348/7c717a0e-ec77-41c9-9c4f-83479cb755c4">
](url)

**Comment by Goldziher** [comment]
I think this can be filed as a bug when using `pgx`

**Comment by andrewmbenton** [comment]
I see, thanks for updating the example. I found an older issue (https://github.com/sqlc-dev/sqlc/issues/2801) which describes the difference in behavior when using different drivers. There is an override workaround described there that should work for you.

I think arguably `json.RawMessage` should be the default, but since it's a breaking change it will have to wait for a major release. I've added it to the list.
